docs/docs/adrs/adr-015-partial-set-security.md

```diff
@@ -0,0 +1,293 @@
+    totalPower := k.ComputeTotalPower(ctx)
+    topNThreshold := k.GetTopN(ctx, chainId) / 100
+    for _, val := range valUpdates {
+        powerSum = powerSum.Add(sdk.NewDecFromInt(sdk.NewInt(val.Power)))
+        // if powerSum / totalPower > topNThreshold
+        if k.IsTopN(ctx, chainId) && powerSum.Quo(totalPower).GT(topNThreshold) {
+            // is Top N validator
+            updates = append(updates, abci.ValidatorUpdate{
+                PubKey: val.PubKey,
+                Power:  val.Power,
+            })
+       } else {
+       if k.IsValidatorOptedIn(ctx, chainId, val.PubKey) {
+            updates = append(updates, abci.ValidatorUpdate{
+                PubKey: val.PubKey,
+                Power:  val.Power,
+            })
+        } else {
+            // remove this validator
+            updates = append(updates, abci.ValidatorUpdate{
+                PubKey: val.PubKey,
+                Power:  0,
+            })
+        }
+    }
+
+    return updates
+}
+```
+And then we change `QueueVSCPackets` to:
+
+```
+// QueueVSCPackets queues latest validator updates for every registered consumer chain
+func (k Keeper) QueueVSCPackets(ctx sdk.Context) {
+    valUpdateID := k.GetValidatorSetUpdateId(ctx) // current valset update ID
+    // Get the validator updates from the staking module.
+    // Note: GetValidatorUpdates panics if the updates provided by the x/staking module
+    // of cosmos-sdk is invalid.
+    valUpdates := k.stakingKeeper.GetValidatorUpdates(ctx)
+
+    for _, chain := range k.GetAllConsumerChains(ctx) {
+        // Apply the key assignment to the validator updates.
+        valUpdates := k.GetPartialSetValidators(ctx, chain.ChainId, valUpdates)
+        valUpdates := k.MustApplyKeyAssignmentToValUpdates(ctx, chain.ChainId, valUpdates)
+        ...
+```
+
+Note that we use `0` to remove validators that are not opted in as described [here](https://docs.cometbft.com/v0.37/spec/abci/abci++_methods#endblock) and as is done for the [assignment of consumer keys](https://github.com/cosmos/interchain-security/blob/v3.3.0/x/ccv/provider/keeper/key_assignment.go#L525).
+
+### How do we distribute rewards?
+Currently, rewards are distributed as follows: The consumer [periodically sends rewards](https://github.com/cosmos/interchain-security/blob/v3.3.0/x/ccv/consumer/keeper/distribution.go#L148) on the provider `ConsumerRewardsPool` address. The provider then [transfers those rewards to the fee collector address](https://github.com/cosmos/interchain-security/blob/v3.3.0/x/ccv/provider/keeper/distribution.go#L77) and those transferred rewards are distributed to validators and delegators.
+
+In PSS, we distribute rewards only to validators that actually validated the consumer chain. To do this we generate a slice with bogus `VoteInfo` that contains the opted in (or Top N) validators and call [AllocateTokens](https://github.com/cosmos/cosmos-sdk/blob/v0.47.7/x/distribution/keeper/allocation.go#L14).
```

Comment:
Why not use AllocateTokensToValidator https://github.com/cosmos/cosmos-sdk/blob/v0.47.7/x/distribution/keeper/allocation.go#L68?

Reply:
Doesn't seem like AllocateTokens will work because it is pulling out of the distribution at the top level instead of taking tokens as an input

Reply:
Resolving this since we removed the exact reward distribution computation from the ADR for now.